pep-0697.rst

```diff
@@ -0,0 +1,357 @@
+versions/implementations of the interpreter.)
+
+However, when a particular type's instance struct is hidden, it becomes
+difficult to subclass it.
+The usual subclassing pattern, explained `in the tutorial <https://docs.python.org/3/extending/newtypes_tutorial.html#subclassing-other-types>`_,
+is to put the base class ``struct`` as the first member of the subclass ``struct``.
+The tutorial shows this on a ``list`` subtype with extra state; adapted to
+a heap type (``PyType_Spec``) the example reads:
+
+.. code-block:: c
+
+    typedef struct {
+        PyListObject list;
+        int state;
+    } SubListObject;
+
+    static PyType_Spec Sublist_spec = {
+        .name = "sublist.SubList",
+        .basicsize = sizeof(SubListObject),
+        .itemsize = 0,
+        .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+        .slots = SubList_slots
+    };
+
+Since the superclass struct (``PyListObject``) is part of the subclass struct
+(``SubListObject``):
+
+- ``PyListObject`` size must be known at compile time, and
+- the size must be the same across all interpreters/versions the compiled
+  extension is ABI-compatible with.
+
+But in limited API/stable ABI, we do not expose the size of ``PyListObject``,
+so that it can vary between CPython versions (and even between possible
+alternate ABI-compatible C API implementations).
+
+With the size not available, limited API users must resort to workarounds such
+as querying ``__basicsize__`` and plugging it into ``PyType_Spec`` at runtime,
+and divining the correct offset for their extra data.
+This requires making assumptions about memory layout, which the limited API
+is supposed to hide.
+
+
+Extending variable-size objects
+-------------------------------
+
+Another scenario where the traditional way to extend an object does not work
+is variable-sized objects, i.e. ones with non-zero ``tp_itemsize``.
+If the instance struct ends with a variable-length array (such as
+in ``tuple`` or ``int``), subclasses cannot add their own extra data without
+detailed knowledge about how the superclass allocates and uses its memory.
+
+Some types, such as CPython's ``PyHeapType``, handle this by storing
+variable-sized data after the fixed-size struct.
+This means that any subclass can add its own fixed-size data.
+(Only one class in the inheritance hierarchy can use variable-sized data, though.)
+This PEP proposes API that makes this practice easier, and ensures the
+variable-sized data is properly aligned.
+
+Note that many variable-size types, like ``int`` or ``tuple``, do not use
+this mechanism. 
+This PEP does not propose any changes to existing variable-size types (like
+``int`` or ``tuple``) except ``PyHeapType``.
+
+
+Extending ``PyHeapType`` specifically
+-------------------------------------
+
+The motivating problem this PEP solves is creating metaclasses, that is,
+subclasses of ``type``.
+The underlying ``PyHeapTypeObject`` struct is both and variable-sized, and
```

Comment:
both _what_ and variable-sized? Missing word?

Reply:
```suggestion
The underlying ``PyHeapTypeObject`` struct is both variable-sized and
```
Word *should* be missing. Thanks for the catch!